### PR TITLE
Refactor ann and unet

### PR DESCRIPTION
**Pull request by amangupta2**
This is an extension of PR #34 and #30 

No work has been done wrt #30 yet. However, all suggestions in #34 have now been implemented for both era5_training and ifs_transfer_learning.

Another folder era5_training is created to replace the two folders ann_cnn_training and attention_net. All the supporting function definitions have been moved to the newly created utils folder. The code is not fully ready yet as I am still working on merging the training for both the models into a single file.

### CODE REVIEW

File: era5_training/training.py
Comment (TomMelt) [inline era5_training/training.py:1]: @TomMelt to review this

Comment (amangupta2) [reply era5_training/training.py:1]: Have made the following changes so far:

1) Merged ANN and UNet training on ERA5 into a single file
2) Merged ANN and UNet inference into a single file, and also inference on ERA5 and IFS into a single file. This dual training on ERA5 and IFS is only available for models trained on ERA5, i.e., this script should not be used for inference on transfer learning models
3) Deleted all the obsolete files. Now the code is divided into three folders only: 
---- /utils: which contains all the necessary subroutines
---- /era5_training: stage 1 training on ERA5 and inference on ERA5 and IFS validation set. This for both ANN and UNets
--- /ifs_transfer_learning: this is stage 2 training to re-train parts of ERA5 trained models on high-resolution IFS data. This directory also contains two different inference files for ERA5 and IFS. I have no plans to merge them into a single file as of now, but I am open to doing it. We can discuss this.

The following tasks remain:
a) #30 Creating a generalized ANN class for ablation studies
b) #33 Minor changes to quick_inference_script.py

Comment (amangupta2) [reply era5_training/training.py:1]: Also, added code snippet in the quick_inference_script.py to download model weights from the new hugging face repository to scratch

File: quick_inference_script/two_sample_inference.py
Comment (TomMelt) [inline quick_inference_script/two_sample_inference.py:1]: we currently import `argparse` but it isn't used. In this file it would also be handy. I will look at adding it tomorrow when I test the inference script.

File: quick_inference_script/two_sample_inference.py [hunk 1]
```diff
@@ -3,114 +3,113 @@
...
 import sys
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
-import torch.optim as optim
+# import torch.optim as optim
+
+# -----------------------------------------------------------
+import logging
+import argparse
```
Comment (TomMelt) [inline quick_inference_script/two_sample_inference.py:19]: we currently import `argparse` but it isn't used. In this file it would also be handy. I will look at adding it tomorrow when I test the inference script.

File: quick_inference_script/two_sample_inference.py [hunk 2]
```diff
@@ -3,114 +3,113 @@
...
+        shell=True,
+    )
+# OR
+# curl -L -O --output-dir /glade/derecho/scratch/agupta/hugging_face_checkpoints https://huggingface.co/amangupta2/iccs_coupling_checkpoints/resolve/main/ann_cnn_1x1_global_global_era5_uvthetaw__train_epoch94.pt
+# curl -L -O --output-dir /glade/derecho/scratch/agupta/hugging_face_checkpoints https://huggingface.co/amangupta2/iccs_coupling_checkpoints/resolve/main/attnunet_era5_global_global_uvthetaw_mseloss_train_epoch119.pt
 
 # assumes ckpts stored in model_ckpt dir
 if model == "attention":
-    PATH = f"model_ckpt/attnunet_era5_{domain}_{vertical}_{features}_mseloss_train_epoch{epoch}.pt"
+    PATH = f"/glade/derecho/scratch/agupta/hugging_face_checkpoints/attnunet_era5_{domain}_{vertical}_{features}_mseloss_train_epoch{epoch}.pt"
```
Comment (TomMelt) [inline quick_inference_script/two_sample_inference.py:64]: it will help me tidy these bits so it runs on my laptop

**Review by amangupta2 — COMMENTED** [review]
@TomMelt Yes, this looks good. These files have now been moved to ../utils/ so they can be deleted. I somehow missed these during the last commit.

File: era5_training/inference.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+bs_train = 40  # 80 (80 works for most). (does not work for global uvthetaw)
+bs_test = bs_train
+
+# --------------------------------------------------
+model = args.model
+domain = args.horizontal
+vertical = args.vertical
+features = args.features  # sys.argv[2]  #'uvthetaw' # 'uvtheta', ''uvthetaw', or 'uvw' for troposphere | additionally 'uvthetaN2' and 'uvthetawN2' for stratosphere_only
+dropout = 0
```
Comment (TomMelt) [inline era5_training/inference.py:113]:
```suggestion
dropout = 0.0
```

File: utils/dataloader_definition.py [hunk 4]
```diff
@@ -283,6 +283,8 @@ def __init__(self, files, domain, vertical, manual_shuffle, features, region="1a
         self.features = features
         self.inp = self.ds["features"]
         self.out = self.ds["output"]
+        self.lat = self.ds["lat"]
+        self.lon = self.ds["lon"]
```
Comment (TomMelt) [inline utils/dataloader_definition.py:287]: @amangupta2 , I had to add these to get the UNet inference working with `inference.py`. Should these have been included?

File: era5_training/inference.py (hunk 3 above)
Comment (TomMelt) [reply era5_training/inference.py:113]: this fixes an annoying bug when loading `torchscript` models. It complains that dropout is an integer when it expects a float if we don't explicitly specify `0.0`

File: quick_inference_script/two_sample_inference.py [hunk 5]
```diff
@@ -3,114 +3,113 @@
...
     model = model.to(device)
     model.eval()
 
-out = f"{model}_output.npz"
-if model == "ann":
-    Inference_and_Save_ANN_CNN(model, testset, testloader, bs_test, device, stencil, out)
-elif model == "attention":
-    Inference_and_Save_AttentionUNet(model, testset, testloader, bs_test, device, out)
+# Load input into INP - single batch - as we would have from CAM
+OUT = model(INP)
```
Comment (TomMelt) [inline quick_inference_script/two_sample_inference.py:112]: `INP` has not been initialized in this code

Comment (TomMelt) [reply quick_inference_script/two_sample_inference.py:112]: this file will be deleted as it is no longer needed

**Review by TomMelt — APPROVED** [review]
This PR is ready to go. It has greatly simplified the code maintenance and it has been tested with the mwe branch for fortran and python
